modules/seo-tools/jetpack-seo-titles.php

```diff
@@ -0,0 +1,255 @@
+ * Possible values for 'value' are: any string in case that 'type' is set
+ * to 'string', or allowed token values for page type in case that 'type'
+ * is set to 'token'.
+ *
+ * Examples of valid formats:
+ *
+ * [
+ *  'front_page' => [
+ *      [ 'type' => 'string', 'value' => 'Front page title and site name:'],
+ *      [ 'type' => 'token', 'value' => 'site_name']
+ *  ],
+ *  'posts' => [
+ *      [ 'type' => 'token', 'value' => 'site_name' ],
+ *      [ 'type' => 'string', 'value' => ' | ' ],
+ *      [ 'type' => 'token', 'value' => 'post_title' ]
+ *  ],
+ *  'pages' => [],
+ *  'groups' => [],
+ *  'archives' => []
+ * ]
+ *  Custom title for given page type is created by concatenating all of the array 'value' parts.
+ *  Tokens are replaced with their corresponding values for current site.
+ *  Empty array signals that we are not overriding the default title for particular page type.
+ */
+
+/**
+ * Class containing utility static methods for managing SEO custom title formats.
+ */
+class Jetpack_SEO_Titles {
+	/**
+	 * Site option name used to store custom title formats.
+	 */
+	const TITLE_FORMATS_OPTION = 'advanced_seo_title_formats';
+
+	/**
+	 * Retrieves custom title formats from site option.
+	 *
+	 * @return array Array of custom title formats, or empty array.
+	 */
+	public static function get_custom_title_formats() {
+		if( Jetpack_SEO_Utils::is_enabled_jetpack_seo() ) {
+			return get_option( self::TITLE_FORMATS_OPTION, array() );
+		}
+
+		return array();
+	}
+
+	/**
+	 * Returns tokens that are currently supported for each page type.
+	 *
+	 * @return array Array of allowed token strings.
+	 */
+	public static function get_allowed_tokens() {
+		return array(
+			'front_page' => [ 'site_name', 'tagline' ],
+			'posts'      => [ 'site_name', 'tagline', 'post_title' ],
+			'pages'      => [ 'site_name', 'tagline', 'page_title' ],
+			'groups'     => [ 'site_name', 'tagline', 'group_title' ],
+			'archives'   => [ 'site_name', 'tagline', 'date' ],
+		);
+	}
+
+	/**
+	 * Used to modify the default title with custom SEO title.
+	 *
+	 * @param string $default_title Default title for current page.
+	 *
+	 * @return string Custom title with replaced tokens or default title.
+	 */
+	public static function get_custom_title( $default_title = '' ) {
+		// Don't filter title for unsupported themes.
+		if ( self::is_conflicted_theme() ) {
+			return $default_title;
+		}
+
+		$page_type = self::get_page_type();
+
+		// Keep default title if invalid page type is supplied.
+		if ( empty( $page_type ) ) {
+			return $default_title;
+		}
+
+		$title_formats = self::get_custom_title_formats();
+
+		// Keep default title if user has not defined custom title for this page type.
+		if ( empty( $title_formats[ $page_type ] ) ) {
+			return $default_title;
+		}
+
+		if ( ! Jetpack_SEO_Utils::is_enabled_jetpack_seo() ) {
+			return $default_title;
+		}
+
+		$custom_title = array_reduce( $title_formats[ $page_type ], function ( $title, $item ) {
```

Comment:
Refactored this to avoid using anonymous function in 462db814f9660e49bb00eff953fbc1a4c07a3945.